src/Opaleye/Internal/HaskellDB/PrimQuery.hs

```diff
@@ -60,6 +60,7 @@ data BinOp      = (:==) | (:<) | (:<=) | (:>) | (:>=) | (:<>)
                 | (:->) | (:->>) | (:#>) | (:#>>)
                 | (:@>) | (:<@) | (:?) | (:?|) | (:?&)
                 | (:&&) | (:<<) | (:>>) | (:&<) | (:&>) | (:-|-)
+                | OpArrayIndex
```

Comment:
I don't understand why you've shoehorned this into `BinOp`.  It seems to me that an additional `ArrayIndex PrimExpr PrimExpr` for `PrimExpr` makes more sense.